### PR TITLE
Prefab Bug Fix - Remove unexpected transform update of container entity for add-entity

### PR DESCRIPTION
**Pull request by forkercat**
Signed-off-by: Junhao Wang <wjunhao@amazon.com>

## What does this PR do?

Fix: https://github.com/o3de/o3de/issues/13155

This PR addresses the issue of an unexpected patch for updating the parent container entity's transform in the source template. During add-entity operation, if the container entity is being focused and we fetch from focused template, the transform in the retrieved entity DOM will be empty. An unexpected patch will be generated to update the transform.

- Patch 1: Add new entity DOM
- Patch 2: Update parent's child entity order
- **Patch 3: Update parent entity's transform    <-- unexpected**

In this case, we need to change it to "fetch from root" such that the transform contains data and later patch generation would not generate an extra patch to update parent transform.

## How was this PR tested?

The recently added unit tests could not verify the redo/undo patches. With changes of connecting PrefabTestFixture to EditorRequestBus going in (https://github.com/o3de/o3de/pull/13211), we should be able to correctly capture the behavior of parent update patches and add some new validations for patch types in undo node unit tests.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoAddEntity.cpp [hunk 1]
```diff
@@ -46,21 +47,50 @@ namespace AzToolsFramework
...
+            // not contain transform data seen from root, but after-state entity DOM from serialization always
+            // contain the transform data, which may generate unexpected patches to update transform.
+            if (parentEntityId == focusedInstance.GetContainerEntityId())
             {
-                // DOM value pointers can't be relied upon if the original DOM gets modified after pointer creation.
-                const PrefabDomValue* parentEntityDomInFocusedTemplate = entityPathInFocusedTemplate.Get(focusedTemplateDom);
-                AZ_Assert(parentEntityDomInFocusedTemplate,
-                    "Could not load parent entity's DOM from the focused template's DOM. "
-                    "Focused template id: '%llu'.", static_cast<AZ::u64>(m_templateId));
+                PrefabDom parentContainerEntityDom;
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoAddEntity.cpp:56]: nit: Should this be called parentEntityDomBeforeAddingEntity for consistency?

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoAddEntity.cpp [hunk 2]
```diff
@@ -46,21 +47,50 @@ namespace AzToolsFramework
...
+                m_instanceDomGeneratorInterface->GenerateEntityDom(parentContainerEntityDom, parentEntity);
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoAddEntity.cpp:57]: yeah...the names are definitely confusing. It took me a while to figure out that GenerateDomForEntity and GenerateEntityDom are actually 2 different functions doing 2 different things. Please add a task to rename these

**Review by srikappa-amzn — APPROVED** [review]
Thanks for proactively fixing this!

File: Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoAddEntity.cpp (hunk 2 above)
Comment (forkercat) [reply Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoAddEntity.cpp]: Do you think we still need InstanceDomGenerator? I think these two functions fit better in InstanceToTemplatePropagator. And keeping them near GenerateDomForEntity and GenerateDomForInstance would be good for us to maintain and to be clear about their functionalities imo, or vice versa moving the other two to InstanceDomGenerator.

**Review by AMZN-daimini — APPROVED** [review]
Agree with the direction of adding some unit tests to prevent regressions as a follow up :)

**Pull request merged**